### PR TITLE
[Aqara] Pet Feeder Edge Device Driver

### PR DESCRIPTION
**Pull request by seojune79**

### CODE REVIEW

**Review by greens — CHANGES_REQUESTED** [review]
We'd like the `Aqara` folder here to be at the same level in the directory structure as the `SmartThings` folder.

File: drivers/SmartThings/aqara/src/init.lua [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+      [capabilities.feederPortion.commands.setPortion.NAME] = feedPortion_handler
+    }
+  },
+  zigbee_handlers = {
+    attr = {
+        [PRIVATE_CLUSTER_ID] = {
+            [PRIVATE_ATTR_ID] = petFeeder_handler
+        }
+    }
+  },
```
Comment (greens) [inline drivers/SmartThings/aqara/src/init.lua:158]: nit: spacing here is off

File: drivers/SmartThings/aqara/src/init.lua [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+local function do_payload(device, funcA, funcB, funcC, op_code, length, value)
+  local seq_num = 1
+  if device:get_field(SEQ_NUM) ~= nil and device:get_field(SEQ_NUM) < 255 then
+    seq_num = device:get_field(SEQ_NUM)+1
+  end
+  local data = "\x00"..string.char(0xFF&op_code)..string.char(0xFF&seq_num)..string.char(0xFF&funcA)
+    ..string.char(0xFF&funcB)..string.char(0xFF&(funcC>>8))..string.char(0xFF&funcC)..string.char(0xFF&length)
+  for i=length-1, 0, -1 do
+    local tmp = 0xFF&(value>>(i*8))
+      data = data..string.char(tmp)
```
Comment (greens) [inline drivers/SmartThings/aqara/src/init.lua:41]: nit: spacing here

**Comment by greens** [comment]
It looks like Github Actions ran into some issue running the linter and uploading the driver and didn't maintain the logs of the errors. Hopefully your changes will cause them to run again (successfully).

File: drivers/SmartThings/aqara/src/init.lua (hunk 1 above)
Comment (seojune79) [reply drivers/SmartThings/aqara/src/init.lua:158]: OK, I will update the code.

File: drivers/SmartThings/aqara/src/init.lua [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    added = device_added,
+    infoChanged = device_info_changed,
+    doConfigure = device_configure
+  },
+  can_handle = function(opts, driver, device, ...)
+    return device:get_model() == "aqara.feeder.acn001"
+  end
+}
+
+defaults.register_for_default_handlers(aqara_pet_feeder_handler, aqara_pet_feeder_handler.supported_capabilities)
```
[the remark this answers is not in the document]
Comment (seojune79) [reply drivers/SmartThings/aqara/src/init.lua:169]: I referred to the zigbee-button init code.
https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/main/drivers/SmartThings/zigbee-button/src/init.lua#L107

Could you tell me how to edit it?

**Comment by seojune79** [comment]
> We'd like the `Aqara` folder here to be at the same level in the directory structure as the `SmartThings` folder.

OK, I will change the folder according to your guide.

File: drivers/SmartThings/aqara/src/init.lua (hunk 2 above)
Comment (seojune79) [reply drivers/SmartThings/aqara/src/init.lua:41]: OK, I will edit the code.

**Comment by greens** [comment]
Just in case you can't see them, here are your luacheck failures:

```
Checking drivers/SmartThings/aqara/src/init.lua   2 warnings

    drivers/SmartThings/aqara/src/init.lua:43:120: (W[6](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/actions/runs/4320165114/jobs/7593682112#step:4:7)12) line contains trailing whitespace
    drivers/SmartThings/aqara/src/init.lua:96:13: (W211) unused variable waiting_timer

Checking drivers/SmartThings/aqara/src/test/test_aqara_pet_feeder.lua [7](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/actions/runs/4320165114/jobs/7593682112#step:4:8) warnings

    drivers/SmartThings/aqara/src/test/test_aqara_pet_feeder.lua:54:55: (W612) line contains trailing whitespace
    drivers/SmartThings/aqara/src/test/test_aqara_pet_feeder.lua:71:119: (W612) line contains trailing whitespace
    drivers/SmartThings/aqara/src/test/test_aqara_pet_feeder.lua:[8](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/actions/runs/4320165114/jobs/7593682112#step:4:9)8:11[9](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/actions/runs/4320165114/jobs/7593682112#step:4:10): (W612) line contains trailing whitespace
    drivers/SmartThings/aqara/src/test/test_aqara_pet_feeder.lua:96:[11](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/actions/runs/4320165114/jobs/7593682112#step:4:12)9: (W6[12](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/actions/runs/4320165114/jobs/7593682112#step:4:13)) line contains trailing whitespace
    drivers/SmartThings/aqara/src/test/test_aqara_pet_feeder.lua:[14](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/actions/runs/4320165114/jobs/7593682112#step:4:15)2:11: (W211) unused function reset_status
    drivers/SmartThings/aqara/src/test/test_aqara_pet_feeder.lua:[15](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/actions/runs/4320165114/jobs/7593682112#step:4:16)5:119: (W612) line contains trailing whitespace
    drivers/SmartThings/aqara/src/test/test_aqara_pet_feeder.lua:196:119: (W612) line contains trailing whitespace
```

**Comment by greens** [comment]
please see my note about `git mv` ing these to an Aqara directory on the same level as Smartthings

**Comment by dljsjr** [comment]
The `Aqara` directory should be a landing place for this and for all future Aqara drivers, so we should probably not have the driver source code files stored at the top level of the manufacturer directory. The new Aqara folder is in the correct place, but the driver code should probably be in a `pet-feeder` subdirectory or something. You can see how the `SmartThings` directory contains several drivers in it as subdirectories, we should do that with Aqara as well.

**Review by dljsjr — CHANGES_REQUESTED** [review]
The `Aqara` directory should be a landing place for this and for all future Aqara drivers, so we should probably not have the driver source code files stored at the top level of the manufacturer directory. The new Aqara folder is in the correct place, but the driver code should probably be in a `pet-feeder` subdirectory or something. You can see how the `SmartThings` directory contains several drivers in it as subdirectories, we should do that with Aqara as well.

**Comment by greens** [comment]
```
Checking drivers/Aqara/src/init.lua               2 warnings

    drivers/Aqara/src/init.lua:2:7: (W211) unused variable defaults
    drivers/Aqara/src/init.lua:100:13: (W211) unused variable waiting_timer

Checking drivers/Aqara/src/test/test_aqara_pet_feeder.lua 7 warnings

    drivers/Aqara/src/test/test_aqara_pet_feeder.lua:54:55: (W[6](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/actions/runs/4349927260/jobs/7616998315#step:4:7)12) line contains trailing whitespace
    drivers/Aqara/src/test/test_aqara_pet_feeder.lua:[7](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/actions/runs/4349927260/jobs/7616998315#step:4:8)1:119: (W612) line contains trailing whitespace
    drivers/Aqara/src/test/test_aqara_pet_feeder.lua:[8](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/actions/runs/4349927260/jobs/7616998315#step:4:9)8:11[9](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/actions/runs/4349927260/jobs/7616998315#step:4:10): (W612) line contains trailing whitespace
    drivers/Aqara/src/test/test_aqara_pet_feeder.lua:96:[11](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/actions/runs/4349927260/jobs/7616998315#step:4:12)9: (W6[12](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/actions/runs/4349927260/jobs/7616998315#step:4:13)) line contains trailing whitespace
    drivers/Aqara/src/test/test_aqara_pet_feeder.lua:[14](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/actions/runs/4349927260/jobs/7616998315#step:4:15)2:11: (W211) unused function reset_status
    drivers/Aqara/src/test/test_aqara_pet_feeder.lua:[15](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/actions/runs/4349927260/jobs/7616998315#step:4:16)5:119: (W612) line contains trailing whitespace
    drivers/Aqara/src/test/test_aqara_pet_feeder.lua:196:119: (W612) line contains trailing whitespace
```

**Comment by seojune79** [comment]
> subdirectories

I have changed the directory structure.

**Review by greens — APPROVED** [review]

**Review by dljsjr — APPROVED** [review]

**Review by tpmanley — APPROVED** [review]
I've been using this driver for a few days and it seems to be working well

**Pull request merged**